### ISSUE TITLE
Invalid SQL generated for SQL server when grouping by complex (calculated) properties

### ISSUE THREAD

**Issue by EuroNode** [dotnet/efcore#26654]
I have run into an issue with generated SQL for SQL server when grouping by complex (calculated) properties.

For example I need to do something like this to gather statistics:

```c#
var result = await _db.Users.Select(iu => new
{
	iu.Id,
	Planned = _db.LifeCorrectionPlans.Where(lcp => lcp.PatientId == iu.Id && lcp.ApprovedOn.HasValue).OrderByDescending(lcp => lcp.ApprovedOn).FirstOrDefault(lcp => lcp.FamilyPhysicianCount.HasValue).FamilyPhysicianCount,
	Finished = _db.TimeSlotBookings.Count(tsb => tsb.UserId == iu.Id && tsb.Finished == true)
})
.Where(idp => idp.Planned > 0)
.Select(idp => new { idp.Id, idp.Planned, idp.Finished })
.GroupBy(idp => new { idp.Planned, idp.Finished }, (key, data) => new { key.Planned, key.Finished, Count = data.Count() })
.ToListAsync();
```

This produces following SQL query:

```sql
SELECT (
    SELECT TOP(1) [l].[FamilyPhysician]
    FROM [LifeCorrectionPlans] AS [l]
    WHERE (([l].[PatientId] = [i].[Id]) AND [l].[ApprovedOn] IS NOT NULL) AND [l].[FamilyPhysician] IS NOT NULL
    ORDER BY [l].[ApprovedOn] DESC) AS [Planned], (
    SELECT COUNT(*)
    FROM [TimeSlotBookings] AS [t]
    WHERE ([t].[UserId] = [i].[Id]) AND ([t].[Finished] = CAST(1 AS bit))) AS [Finished], COUNT(*) AS [Count]
FROM [IdentityUsers] AS [i]
WHERE (
    SELECT TOP(1) [l0].[FamilyPhysician]
    FROM [LifeCorrectionPlans] AS [l0]
    WHERE (([l0].[PatientId] = [i].[Id]) AND [l0].[ApprovedOn] IS NOT NULL) AND [l0].[FamilyPhysician] IS NOT NULL
    ORDER BY [l0].[ApprovedOn] DESC) > 0
GROUP BY (
    SELECT TOP(1) [l].[FamilyPhysician]
    FROM [LifeCorrectionPlans] AS [l]
    WHERE (([l].[PatientId] = [i].[Id]) AND [l].[ApprovedOn] IS NOT NULL) AND [l].[FamilyPhysician] IS NOT NULL
    ORDER BY [l].[ApprovedOn] DESC), (
    SELECT COUNT(*)
    FROM [TimeSlotBookings] AS [t]
    WHERE ([t].[UserId] = [i].[Id]) AND ([t].[Finished] = CAST(1 AS bit)))
```

Although this would be a valid SQL on some DB providers, SQL server does not support group by such complex calculated statements and I get this exception when query is run:
Microsoft.Data.SqlClient.SqlException (0x80131904): Cannot use an aggregate or a subquery in an expression used for the group by list of a GROUP BY clause.

I can't seem to find any workaround which would allow me to run such query via EF short of adding mapping for result set to context, and running raw SQL query which would look very similar, but use a subquery like:

```sql
SELECT [Planned], [Finished], Count(*) AS [Count] FROM
(SELECT (
    SELECT TOP(1) [l].[FamilyPhysician]
    FROM [LifeCorrectionPlans] AS [l]
    WHERE (([l].[PatientId] = [i].[Id]) AND [l].[ApprovedOn] IS NOT NULL) AND [l].[FamilyPhysician] IS NOT NULL
    ORDER BY [l].[ApprovedOn] DESC) AS [Planned], (
    SELECT COUNT(*)
    FROM [TimeSlotBookings] AS [t]
    WHERE ([t].[UserId] = [i].[Id]) AND ([t].[Finished] = CAST(1 AS bit))) AS [Finished]
FROM [IdentityUsers] AS [i]
WHERE (
    SELECT TOP(1) [l0].[FamilyPhysician]
    FROM [LifeCorrectionPlans] AS [l0]
    WHERE (([l0].[PatientId] = [i].[Id]) AND [l0].[ApprovedOn] IS NOT NULL) AND [l0].[FamilyPhysician] IS NOT NULL
    ORDER BY [l0].[ApprovedOn] DESC) > 0)
AS Q
GROUP BY [Planned], [Finished]
```

Tried on:
EF Core version: 5.0.12
Database provider: Microsoft.EntityFrameworkCore.SqlServer
Target framework: NET 5.0

**Comment by roji** [comment]
@EuroNode can you give the recently-released EF Core 6.0 a try? There have been some substantial improvements to GroupBy translation support.

**Comment by EuroNode** [comment]
I wanted to try, but this is part of a quite big solution (with many parts based on .NET standard) and it would be time consuming to port it to .Net 6. I just don't have time for that now.
